shims/spark310/src/main/scala/com/nvidia/spark/rapids/shims/spark310/ParquetCachedBatchSerializer.scala

```diff
@@ -405,87 +431,301 @@ class ParquetCachedBatchSerializer extends CachedBatchSerializer with Arm {
+          requestedSchema.toStructType)
+        val columnarBatch: ColumnarBatch = new ColumnarBatch(columnVectors
+          .asInstanceOf[Array[org.apache.spark.sql.vectorized.ColumnVector]])
+        val missingColumns = new Array[Boolean](reqParquetSchema.getFieldCount)
+        var columnReaders: Array[VectorizedColumnReader] = null
+
+        var rowsReturned: Long = 0L
+        var totalRowCount: Long = 0L
+        var totalCountLoadedSoFar: Long = 0
+        var batchIdx = 0
+        var numBatched = 0
+
+        import scala.collection.JavaConverters._
+        for (block <- parquetFileReader.getRowGroups.asScala) {
+          this.totalRowCount += block.getRowCount
+        }
 
-            val catalystColumns =
-              getColumnIndices(cacheAttributes, cacheAttributes).map { i => "_col" + i }
+        for (i <- missingColumns.indices) {
+          if (missingColumns(i)) {
+            columnVectors(i).putNulls(0, capacity)
+            columnVectors(i).setIsConstant()
+          }
+        }
 
-            val catalystSchema = cacheAttributes.zip(catalystColumns).map {
-              case (attr, name) => attr.withName(name)
+        @throws[IOException]
+        def checkEndOfRowGroup(): Unit = {
+          if (rowsReturned != totalCountLoadedSoFar) return
+          val pages = parquetFileReader.readNextRowGroup
+          if (pages == null) {
+            throw new IOException("expecting more rows but reached last" +
+              " block. Read " + rowsReturned + " out of " + totalRowCount)
+          }
+          val columns = reqParquetSchema.getColumns
+          val types = reqParquetSchema.asGroupType.getFields
+          columnReaders = new Array[VectorizedColumnReader](columns.size)
+          for (i <- 0 until columns.size) {
+            if (!missingColumns(i)) {
+              columnReaders(i) =
+                new VectorizedColumnReader(columns.get(i), types.get(i)
+                  .getOriginalType, pages.getPageReader(columns.get(i)), null /*convertTz*/,
+                  LegacyBehaviorPolicy.CORRECTED.toString, LegacyBehaviorPolicy.EXCEPTION.toString)
             }
+          }
+          totalCountLoadedSoFar += pages.getRowCount
+        }
 
-            val selectedColumns =
-              getColumnIndices(selectedAttributes, cacheAttributes).map { i => "_col" + i }
-
-            val requestedSchema = selectedAttributes.zip(selectedColumns).map {
-              case (attr, name) => attr.withName(name)
+        def nextBatch: Boolean = {
+          for (vector <- columnVectors) {
+            vector.reset()
+          }
+          columnarBatch.setNumRows(0)
+          if (rowsReturned >= totalRowCount) return false
+          checkEndOfRowGroup()
+          val num = Math.min(capacity.toLong, totalCountLoadedSoFar - rowsReturned).toInt
+          for (i <- columnReaders.indices) {
+            if (columnReaders(i) != null) {
+              val readBatchMethod =
```

Comment:
I have moved this up at the class level so we don't have to look it up every time. Update coming soon